tests/initcoder_test.go

```diff
@@ -0,0 +1,156 @@
+package tests
+
+import (
+	"context"
+	"math/big"
+	"testing"
+
+	"github.com/Fantom-foundation/go-opera/tests/contracts/initcode"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/stretchr/testify/require"
+)
+
+// this constant comes from  https://eips.ethereum.org/EIPS/eip-3860#parameters
+const MAX_INIT_CODE_SIZE uint64 = 49152
+const sufficientGas = uint64(100_000)
+
+func TestInitCodeSizeLimitAndMetered(t *testing.T) {
+	require := require.New(t)
+
+	net, err := StartIntegrationTestNet(t.TempDir())
+	require.NoError(err)
+	defer net.Stop()
+
+	// Deploy the invalid start contract.
+	contract, receipt, err := DeployContract(net, initcode.DeployInitcode)
+	require.NoError(err)
+	require.Equal(types.ReceiptStatusSuccessful, receipt.Status, "failed to deploy contract")
+
+	// deploy measureGasAndAssign to get cost of deploying a contract without create.
+	assignCost, gasUsed := createAndCheckContractWithCodeLenAndGas(require, net, contract, contract.MeasureGasAndAssign, 0, 0_000)
+
+	// -- using CREATE instruction
+	const wordCostCreate uint64 = 2
+	var gasForCreate uint64 = 32000 + gasUsed + assignCost
+	t.Run("create", func(t *testing.T) {
+		testForVariant(t, net, contract, receipt, contract.CreateContractWith, gasForCreate, wordCostCreate)
+	})
+
+	// -- using CREATE2 instruction
+	// create2 has extra costs for hashing, this is explained in
+	// https://eips.ethereum.org/EIPS/eip-3860 and reflected in evm.codes calculator.
+	const wordCostCreate2 uint64 = wordCostCreate + 6
+	var gasForCreate2 uint64 = gasForCreate + 44
+	t.Run("create2", func(t *testing.T) {
+		testForVariant(t, net, contract, receipt, contract.Create2ContractWith, gasForCreate2, wordCostCreate2)
+	})
+
+	t.Run("make a transaction that charges for the init code size", func(t *testing.T) {
+		// transactions charge 4 gas for each zero byte in data.
+		const zeroByteCost uint64 = 4
+		// createa a transaction with 1 word of code.
+		receipt1, err := runTransactionWithCodeSizeAndGas(require, net, 1, sufficientGas)
+		require.NoError(err)
+		require.Equal(types.ReceiptStatusSuccessful, receipt1.Status, "failed on transfer to empty receiver with valid code")
+
+		// createa a transaction with 2 byte of code.
+		receipt2, err := runTransactionWithCodeSizeAndGas(require, net, 2, sufficientGas)
+		require.NoError(err)
+		require.Equal(types.ReceiptStatusSuccessful, receipt2.Status, "failed on transfer to empty receiver with valid code")
+
+		difference := receipt2.GasUsed - receipt1.GasUsed
+		require.Equal(difference, zeroByteCost, "gas difference between 1 and 2 words should be 4, instead got %d",
+			difference)
+	})
+
+	t.Run("make a transaction without enough gas for init code size", func(t *testing.T) {
```

Comment:
I don't understand this one. 
```
	transaction, err := types.SignTx(types.NewTx(&types.AccessListTx{
		ChainID:  chainId,
		Gas:      gas,
		GasPrice: price,
		To:       nil,
		Nonce:    nonce,
		Data:     make([]byte, codeSize),
```

This creates a transaction without code. and allocates a buffer for the data (input for the contract to be executed). It makes sense that it fails because of not having enough gas, but I do not understand what does this have to do with CREATE/CREATE2 
Can you develop?

Reply:
this case is, as you mention, unrealted to CREATE/CREATE2. it is only testing that in a transaction where the init code is less than `MAX_INIT_CODE_SIZE`, it is charged for the of it.

Reply:
I get it now, the mentioned "create transaction". thanks it makes more sense now.

Reply:
as discussed offline, it is clear what this does not.